### DISCUSSION TITLE
Multispeed Random walker

### DISCUSSION THREAD

**Discussion by ygk2zhang** [PMEAL/pytrax#23] [Ideas]
Hi,

We started using Pytrax to study diffusion in porous media and it works great to provide a rough estimate.
Is it possible to implement multispeed random walkers? Currently, images are binary **0** and **1**, where **1** are accessible pixels by random walkers. Suppose we alter this binary system to **0**, **1** and **2**, where random walkers will travel at predefined different speeds for **1** and **2**. The idea is **1** and **2** would represent different pore size ranges which affects the speed of random walkers in certain systems. I would like to know if this would be a lot of work to implement.

Thank you very much,

**Comment by TomTranter** [comment]
Hey @ygk2zhang, nice to hear you are finding the package useful. I thought about having something similar which I think would give you the same results by introducing a jump probability that depends on the greyscale value between 0 and 1. Could have different relationships between grey scale and probability or just linear. I have this implemented in a fork but never got round to merging it. There were a few papers I had found on composite materials that did similar things for thermal conductivity etc. I think it should work fairly well for two three composites or phases but when I applied it to the whole continuum of values that you start getting from tomography images the very low values near 0 ended up trapping the walkers and leading to much higher tortuosity calculations than if they had never been allowed to enter the voxel in the first place. Does that make sense? If you'd like to have a look at what I got up to then I'll push the code to here as I will probably not continue working on the problem myself and am not focused on publishing papers right now. Any thoughts @jgostick ?

**Reply by ygk2zhang** [reply]
Thank you for getting back to me so quickly. I emailed Dr. Gostick first because he is the corresponding author in the Pytrax publication, and told me to write it here. 

 If I understood you correctly, in this greyscale fork must assign values between 0 and 1 to the pixels. Pixels with 1s will have the highest speed or jump probability, and close to 0 will be trapped. How does the jump probability affect the speed- if we have a line of 1s and 0.5s, will the random walkers travel twice as fast in 1s vs 0.5s?

**Reply by TomTranter** [reply]
Yes that's basically it if you have a linear relationship between greyscale and jump probability. It probably better to manipulate greyscale images so that you have banded thresholds. Your original idea would work if you set your fastest speed to 1 and then other phases normalized by the fastest speed. Another way to implement it would be to assume different timesteps for walkers in different phases but then you would have to let all the walkers complete a different number of steps to finish the simulation for a consistent amount of time

**Reply by TomTranter** [reply]
Or you could let them jump continuous distances and not constrain the walkers to pixel centers but I think @jgostick had issues implementing that a while back for some mean free path calculation

**Reply by TomTranter** [reply]
In the interest of open-ness there are also other packages out there that let you do this type of stuff. TauFactor being one where you can definitely set different diffusivity values to multiple phases and possibly another tool from the PMEAL group too

**Comment by ygk2zhang** [comment]
> Yes that's basically it if you have a linear relationship between greyscale and jump probability. It probably better to manipulate greyscale images so that you have banded thresholds. Your original idea would work if you set your fastest speed to 1 and then other phases normalized by the fastest speed. Another way to implement it would be to assume different timesteps for walkers in different phases but then you would have to let all the walkers complete a different number of steps to finish the simulation for a consistent amount of time

Thank you very much for being so helpful,  Dr. Tranter. I think this is what I need. With this, I can set small pores to have the values of 1 and larger pores which will have slower diffusion, will have smaller values. One thing is not clear to me is how to establish the relationship between greyscale values and jump probability in the greyscale Pytrax.

**Comment by TomTranter** [comment]
I created a develop branch here with my changes on my fork merged in. Can you check it out? There are some scripts giving examples. I need to probably tidy up a bit and also get CI and testing running again on this repo. It would be good if you could test this code yourself and suggest/make any changes. I also have a half finished draft paper on this that might be worth looking at if you're interested in finishing it off or incorporating bits. Email is t.g.tranter@gmail.com. Drop me a message if you like
